create-blog-keep-OpenAPI-dry (#1429)
stoplightio/stoplight.io: netlify/blog-posts/keeping-openapi-dry-and-portable.md

---
path: /blog/keeping-OpenAPI-DRY-and-portable
tags:
  - blog
  - blog-featured
  - blog-general
  - blog-design
relatedTags:
  - blog-featured
  - blog-general
  - blog-design
publishedDate: 2020-03-11
author: Phil Sturgeon
title: >-
  Keeping OpenAPI DRY and Portable
subtitle: >-
  Bundling, dereferencing, and how to use them in Stoplight Studio and CLI workflows.
image: /images/keep-open-api-dry.png
includeToc: false
actionBar:
  enabled: true
meta:
  url: https://stoplight.io/blog/keeping-OpenAPI-DRY-and-portable/
  description: >-
    Bundling, dereferencing, and how to use them in Stoplight Studio and CLI workflows.
  favicon: /images/mark_light_bg.png
  robots: 'index, follow'
  title: >-
    Keeping OpenAPI DRY and Portable
  image: /images/keep-open-api-dry.png
  twitter:
    title: >-
      Stoplight Blog: Keeping OpenAPI DRY and Portable
    image: /images/keep-open-api-dry.png
    username: '@stoplightio'
---

> If you mention OpenAPI (or use it's old name "Swagger") to an API developer, there's a chance they groan or make some other funny noise. When you ask why, you'll often hear a terrible tale of 10,000 line YAML files, repeated content, and never-ending git conflicts. Problems like these have been a thing of the past for a while, thanks to editors like [Stoplight Studio](https://stoplight.io/studio) making it easier to work with $ ref's and split files into manageable chunks. However, to transport these separated OpenAPI documents, you'll need to know two terms: bundling and dereferencing!

Reviewer: ```suggestion
If you mention OpenAPI (or use its old name "Swagger") to an API developer, there's a chance they groan or make some other funny noise. When you ask why, you'll often hear a terrible tale of 10,000 line YAML files, repeated content, and never-ending git conflicts. Problems like these have been a thing of the past for a while, thanks to editors like [Stoplight Studio](https://stoplight.io/studio) making it easier to work with $ refs and split files into manageable chunks. However, to transport these separated OpenAPI documents, you'll need to know two terms: bundling and dereferencing!
```

```yaml
Person:
    $ref: "../schemas/person.yaml"
```

The downside to splitting OpenAPI documents up is that many tools don't understand how to read these files. Some tools ask you to upload a single `openapi.yaml` document, and when you upload that single document to the web server, their code has no idea what a `../schemas/person.yaml` is.

Some tooling simply refuses to follow a $ref to an "external file" because it could be a security issue. Others support an external file if it's on the file system but fail to follow a URL for a variety of reasons… Stoplight tooling supports external files and external URLs, but your mileage may vary in the greater world of [OpenAPI tooling](https://openapi.tools/).

> Solutions to this are many; some tools allow you to upload a ZIP file, at which point they sniff all the file contents to figure out which is the "main file", then support $ ref's to other documents. This functionality is pretty rare, a lot of the time tools just break.

Reviewer: ```suggestion
Solutions to this are many; some tools allow you to upload a ZIP file, at which point they sniff all the file contents to figure out which is the "main file", then support $ refs to other documents. This functionality is pretty rare, a lot of the time tools just break.
```

                	example: 57eea03c72381f86e05c35d2
                	type: string
              	status:
                	description: HTTP Response Code
                	example: 200
                	format: int32
                	type: integer
          	pagination:
            	type: object
            	description: |
              	The Pagination Object contains information relating to the number of total results available as well as the number of results fetched and their relative positions.
            	properties:
              	count:
                	description: Total number of items returned.
                	example: 25
                	format: int32
                	type: integer
              	offset:
                	description: Position in pagination.
                	example: 75
                	format: int32
                	type: integer
              	total_count:
                	description: Total number of items available.
                	example: 250
                	format: int32
                	type: integer
        	type: object
    	'400':
      	$ref: '#/responses/BadRequest'
```

> The output here is a little funny lookin at first, but what it's doing is rather smart. It loops through all the $ ref's, and the first time it finds one, it'll replace it with the contents of the $ref target, then the next time it finds one, it'll just change the $ref to target that first definition. In this example, it put `meta` and `pagination` directly into this operation, but the data has a $ref to elsewhere.

Reviewer: ```suggestion
The output here is a little funny looking at first, but what it's doing is rather smart. It loops through all the $ refs, and the first time it finds one, it'll replace it with the contents of the $ref target, then the next time it finds one, it'll just change the $ref to target that first definition. In this example, it put `meta` and `pagination` directly into this operation, but the data has a $ref to elsewhere.
```

```yaml
              data:
                items:
                  $ref: '#/paths/~1gifs~1random/get/responses/200/schema/properties/data'
            	type: array
```

## Bundling Outside of Studio

We added this functionality to Studio using a NodeJS module called [APIDevTools/json-schema-ref-parser](https://github.com/APIDevTools/json-schema-ref-parser). This battle-hardened module handles all $ref logic in Studio, Docs, Prism, and everything else in the Stoplight ecosystem.

Until recently, we maintained a similar NodeJS module [stoplightio/json-ref-parser](https://github.com/stoplightio/json-ref-resolver), but we've combined efforts with this one so we can all solve more significant, more interesting problems, quicker.

Another benefit is that it has a [CLI wrapper swagger-cli](https://github.com/APIDevTools/swagger-cli) available. If you're using another OpenAPI editor, are muddling through with Vim, or need to bundle as part of an automated workflow via the CLI, you can use swagger-cli on the command line to get the same result as if you clicked the button in Studio.

## No $ref Support At All?

> Bundling helps when tools only support some $ ref's, but what if the tool cannot handle $ref at all? If you come across one of these tools, the best thing to do is to run away and [find a replacement](https://openapi.tools/). If there is no alternative tool, try supporting that open-source project with a pull request (via json-schema-ref-parser or an equivalent [JSON Schema tool](https://json-schema.org/implementations.html) in the appropriate programming language).

Reviewer: ```suggestion
Bundling helps when tools only support some $ refs, but what if the tool cannot handle $ref at all? If you come across one of these tools, the best thing to do is to run away and [find a replacement](https://openapi.tools/). If there is no alternative tool, try supporting that open-source project with a pull request (via json-schema-ref-parser or an equivalent [JSON Schema tool](https://json-schema.org/implementations.html) in the appropriate programming language).
```

> Sometimes you are truly stuck with this tool, and for this scenario, there is dereferencing. When using this strategy, $refs are replaced with their values, as though somebody went through doing copy & paste on each one. No more $ ref's exist in the file/object representation. If you have 10 operations referencing the same model 10 times, you now have 10 copies of that model embedded in different parts of the document.

Reviewer: ```suggestion
Sometimes you are truly stuck with this tool, and for this scenario, there is dereferencing. When using this strategy, $refs are replaced with their values, as though somebody went through doing copy & paste on each one. No more $ refs exist in the file/object representation. If you have 10 operations referencing the same model 10 times, you now have 10 copies of that model embedded in different parts of the document.
```
